internal/provider/providers/hetzner/provider.go

```diff
@@ -0,0 +1,340 @@
+		Host:      p.Host(),
+		Provider:  "<a href=\"https://www.hetzner.com\">Hetzner</a>",
+		IPVersion: p.ipVersion.String(),
+	}
+}
+
+func (p *Provider) setHeaders(request *http.Request) {
+	headers.SetUserAgent(request)
+	headers.SetContentType(request, "application/json")
+	headers.SetAccept(request, "application/json")
+	switch {
+	case p.token != "":
+		headers.SetAuthAPIToken(request, p.token)
+	}
+}
+
+// Obtain domain ID.
+// See https://dns.hetzner.com/api-docs#operation/GetZones.
+func (p *Provider) getRecordID(ctx context.Context, client *http.Client, newIP netip.Addr) (
+	identifier string, upToDate bool, err error) {
+	recordType := constants.A
+	if newIP.Is6() {
+		recordType = constants.AAAA
+	}
+
+	u := url.URL{
+		Scheme: "https",
+		Host:   "dns.hetzner.com",
+		Path:   "/api/v1/records",
+	}
+
+	values := url.Values{}
+	values.Set("zone_id", p.zoneIdentifier)
+	values.Set("name", p.host)
+	values.Set("type", recordType)
+	values.Set("page", "1")
+	values.Set("per_page", "1")
+	u.RawQuery = values.Encode()
+
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
+	if err != nil {
+		return "", false, fmt.Errorf("creating http request: %w", err)
+	}
+	p.setHeaders(request)
+
+	response, err := client.Do(request)
+	if err != nil {
+		return "", false, err
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode != http.StatusOK {
+		return "", false, fmt.Errorf("%w: %d: %s",
+			errors.ErrHTTPStatusNotValid, response.StatusCode, utils.BodyToSingleLine(response.Body))
+	}
+
+	decoder := json.NewDecoder(response.Body)
+	listRecordsResponse := struct {
+		Records []struct {
+			ID    string `json:"id"`
+			Value string `json:"value"`
+		} `json:"records"`
+	}{}
+	err = decoder.Decode(&listRecordsResponse)
+	if err != nil {
+		return "", false, fmt.Errorf("json decoding response body: %w", err)
+	}
+
+	switch {
+	case len(listRecordsResponse.Records) == 0:
+		return "", false, fmt.Errorf("%w", errors.ErrReceivedNoResult)
+	case len(listRecordsResponse.Records) > 1:
+		return "", false, fmt.Errorf("%w: %d instead of 1",
+			errors.ErrResultsCountReceived, len(listRecordsResponse.Records))
+	case listRecordsResponse.Records[0].Value == newIP.String(): // up to date
+		return "", true, nil
+	}
+	return listRecordsResponse.Records[0].ID, false, nil
+}
+
+func (p *Provider) CreateRecord(ctx context.Context, client *http.Client, ip netip.Addr) (recordID string, err error) {
+	recordType := constants.A
+
+	if ip.Is6() {
+		recordType = constants.AAAA
+	}
+
+	u := url.URL{
+		Scheme: "https",
+		Host:   "dns.hetzner.com",
+		Path:   fmt.Sprintf("/api/v1/records"),
+	}
+
+	requestData := struct {
+		Type           string `json:"type"`  // constants.A or constants.AAAA depending on ip address given
+		Name           string `json:"name"`  // DNS record name i.e. example.com
+		Value          string `json:"value"` // ip address
```

Comment:
nit comments are not really needed, since we can refer to the struct fields initialization right below

```suggestion
		Type           string `json:"type"` 
		Name           string `json:"name"` 
		Value          string `json:"value"`
```